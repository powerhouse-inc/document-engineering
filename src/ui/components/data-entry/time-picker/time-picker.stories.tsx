import type { Meta, StoryObj } from '@storybook/react'
import { getDefaultArgTypes, StorybookControlCategory } from '../../../../scalars/lib/storybook-arg-types.js'
import { TimePicker } from './time-picker'

/**
 * The `TimePicker` component provides an input field for selecting times.
 * It supports multiple configuration properties like:
 * - label
 * - description
 * - timeFormat
 * - timeIntervals
 * - timeZone
 * - showTimezoneSelect
 *
 * Features include:
 * - Customizable time format (12/24 hour)
 * - Configurable time intervals
 * - Timezone selection support
 * - Custom placeholder support
 *
 * ## Time Picker Customization
 * The time picker can be customized using the `className` prop which accepts a string of Tailwind classes.
 * Each class should follow the format `[&_.time-picker\\_\\_{element}]:{tailwind-classes}`.
 *
 * Base classes available for customization:
 * - .base-picker__input: Input field container
 * - .base-picker__popover: Popover container
 * - .time-picker__content: Main container for the time picker
 * - .time-picker__period: AM/PM selector section
 * - .time-picker__selector: Time selection area (hours/minutes)
 * - .time-picker__buttons: Container for action buttons (Cancel/OK)
 * - .time-picker__select: Individual time selection buttons
 * - .time-picker__hour: Hour display in 24-hour format
 * - .time-picker__minute: Minute display
 * - .time-picker__separator: Colon separator between hours and minutes
 * - .time-picker__am-pm: AM/PM button in 12-hour format
 *
 * # Example Usage
 * When using Tailwind CSS with custom class names, use `String.raw` to properly escape class names:
 *
 * ```tsx
 * <TimePicker
 *   name="custom-time"
 *   label="Select Time"
 *   className={String.raw`
 *     [&_.time-picker\\_\\_content]:bg-blue-100
 *     [&_.time-picker\\_\\_selector]:[&>div]:bg-green-100
 *     [&_.time-picker\\_\\_period]:[&>button]:text-red-500
 *     [&_.time-picker\\_\\_buttons]:[&>button]:bg-gray-800
 *   `}
 * />
 * ```
 *
 * > **Note:** This component does not have built-in validation. If you need built-in validation
 * > you can use the [Time Picker](?path=/docs/scalars-timepickerfield--readme)
 * > component.
 */

const meta: Meta<typeof TimePicker> = {
  title: 'Data Entry/Time Picker',
  component: TimePicker,
  parameters: {
    layout: 'centered',
    form: {
      resetBehavior: 'unmount',
    },
  },
  tags: ['autodocs'],
  argTypes: {
    ...getDefaultArgTypes(),
    timeFormat: {
      control: {
        type: 'select',
      },
      description: 'The format of the time in the time picker',
      table: {
        type: { summary: 'string' },
        category: StorybookControlCategory.COMPONENT_SPECIFIC,
      },

      options: ['hh:mm a', 'HH:mm'],
      defaultValue: { summary: 'hh:mm a' },
    },
    showTimezoneSelect: {
      control: {
        type: 'boolean',
      },
      description: 'Show timezone select',
      table: {
        category: StorybookControlCategory.COMPONENT_SPECIFIC,
        type: { summary: 'boolean' },
      },
    },
    timeIntervals: {
      description: 'The interval between each time option',
      control: {
        type: 'number',
      },

      type: 'number',
      min: 1,
      max: 60,
      table: {
        category: StorybookControlCategory.COMPONENT_SPECIFIC,
      },

      defaultValue: { summary: 1 },
    },
    timeZone: {
      description: 'The timezone to display in the time picker',
      control: {
        type: 'text',
      },
      table: {
        category: StorybookControlCategory.COMPONENT_SPECIFIC,
        type: { summary: 'string' },
      },
    },
    includeContinent: {
      description: 'Show continent name in the timezone select',
      control: {
        type: 'boolean',
        defaultValue: false,
      },
      table: {
        category: StorybookControlCategory.COMPONENT_SPECIFIC,
        defaultValue: { summary: 'false ' },
        type: { summary: 'boolean' },
      },
    },
    placeholder: {
      description: 'The placeholder text for the time picker',
      table: {
        category: StorybookControlCategory.DEFAULT,
        type: { summary: 'string' },
      },
    },
    viewMode: {
      control: 'select',
      options: ['edition', 'addition', 'removal', 'mixed'],
      defaultValue: 'edition',
      description: 'The mode of the time picker',
      table: {
        type: { summary: 'string' },
        category: StorybookControlCategory.DIFF,
      },
    },
    baseValue: {
      control: 'text',
      description: 'The base value of the time picker',
      table: {
        category: StorybookControlCategory.DIFF,
      },
    },
  },

  args: {
    name: 'time-picker-field',
  },
}

export default meta
type Story = StoryObj<typeof meta>

export const Default: Story = {
  args: {
    name: 'time',
    label: 'Pick a time',
    placeholder: 'HH:mm',
  },
}
export const Disabled: Story = {
  args: {
    name: 'time',
    label: 'Pick a time',
    placeholder: 'HH:mm',
    disabled: true,
  },
}

export const Filled: Story = {
  args: {
    name: 'time',
    label: 'Pick a time',
    value: '12:00 PM',
    placeholder: 'HH:mm',
  },
}

export const WithCustomizedCalendar: Story = {
  args: {
    name: 'time',
    label: 'Pick a time',
    placeholder: 'HH:mm',
    className: String.raw`
      [&.base-picker\\_\\_input]:w-[275px]
      [&.base-picker\\_\\_popover]:w-[275px]
      [&_.time-picker\\_\\_content]:bg-red-300
      [&_.time-picker\\_\\_content]:border-2
      [&_.time-picker\\_\\_content]:border-cyan-400
      [&_.time-picker\\_\\_content]:rounded-lg
      [&_.time-picker\\_\\_content]:p-2
      [&_.time-picker\\_\\_content]:transition-all
      [&_.time-picker\\_\\_content]:duration-300
      [&_.time-picker\\_\\_content]:font-mono
      [&_.time-picker\\_\\_content]:text-lg
      [&_.time-picker\\_\\_content]:border-b
      [&_.time-picker\\_\\_content]:border-cyan-400/30
      [&_.time-picker\\_\\_period]:[&>button]:text-blue-500
      [&_.time-picker\\_\\_period]:[&>button[data-selected=true]]:bg-blue-100
      [&_.time-picker\\_\\_period]:[&>button[data-selected=true]]:rounded-md
      [&_.time-picker\\_\\_period]:[&>button[data-selected=true]]:font-bold
      [&_.time-picker\\_\\_selector]:[&>div]:bg-green-500
      [&_.time-picker\\_\\_buttons]:[&>button]:text-gray-900 
      [&_.time-picker\\_\\_select]:[&>button]:text-gray-900
      [&_.time-picker\\_\\_select]:[&>button]:bg-red-500
      [&_.time-picker\\_\\_select]:[&>button]:rounded-lg
      [&_.time-picker\\_\\_select]:[&>button]:p-2
      [&_.time-picker\\_\\_select]:[&>button]:transition-all
      [&_.time-picker\\_\\_select]:[&>button]:duration-300
    `,
  },
}

export const WithDifferencesAddition: Story = {
  args: {
    label: 'Time difference addition',
    value: '12:00 PM',
    baseValue: '11:00 AM',
    viewMode: 'addition',
  },
}
export const WithDifferencesRemoval: Story = {
  args: {
    label: 'Time difference removal',
    value: '12:00 PM',
    baseValue: '11:00 AM',
    viewMode: 'removal',
  },
}
export const WithDifferencesMixed: Story = {
  args: {
    label: 'Time difference mixed',
    value: '12:00 PM',
    baseValue: '11:00 AM',
    viewMode: 'mixed',
  },
}
