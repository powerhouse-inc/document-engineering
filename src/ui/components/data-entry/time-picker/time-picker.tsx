import { FormDescription } from '../../../../scalars/components/fragments/form-description/index.js'
import { FormGroup } from '../../../../scalars/components/fragments/form-group/form-group.js'
import { FormLabel } from '../../../../scalars/components/fragments/form-label/form-label.js'
import { FormMessageList } from '../../../../scalars/components/fragments/form-message/message-list.js'
import { forwardRef, useId } from 'react'
import type { InputBaseProps, WithDifference } from '../../../../scalars/components/types.js'
import type { InputNumberProps } from '../number-input/types.js'
import type { SelectFieldProps } from '../../../../scalars/components/fragments/select-field/index.js'
import { BasePickerField } from '../date-time-picker/base-picker.js'
import TimePickerContent from './subcomponents/time-picker-content.js'
import type { TimeFieldValue } from './type.js'
import { useTimePicker } from './use-time-picker.js'
import { handleKeyDown } from './utils.js'
import TextInputDiff from '../text-input/text-input-diff.js'

interface TimePickerProps
  extends InputBaseProps<TimeFieldValue>,
    Omit<InputNumberProps, 'value' | 'defaultValue'>,
    WithDifference<TimeFieldValue> {
  label?: string
  id?: string
  name: string
  value?: TimeFieldValue
  defaultValue?: TimeFieldValue
  placeholder?: string
  inputProps?: Omit<
    InputBaseProps<string> & {
      onKeyDown?: (event: React.KeyboardEvent<HTMLInputElement>) => void
    },
    'name' | 'onChange' | 'value' | 'defaultValue' | 'onBlur'
  >
  selectProps?: Omit<SelectFieldProps, 'name' | 'options' | 'selectionIcon'>
  timeFormat?: string
  timeIntervals?: number
  showTimezoneSelect?: boolean
  timeZone?: string
  includeContinent?: boolean
  className?: string
}

const TimePicker = forwardRef<HTMLInputElement, TimePickerProps>(
  (
    {
      label,
      id: propId,
      errors,
      name,
      placeholder,
      value,
      defaultValue,
      onChange,
      onBlur,
      description,
      warnings,
      required,
      disabled,
      inputProps,
      selectProps,
      timeFormat = 'hh:mm a',
      showTimezoneSelect,
      timeIntervals,
      timeZone,
      includeContinent,
      className,
      // diff props,
      viewMode = 'edition',
      baseValue,
    },
    ref
  ) => {
    const {
      selectedHour,
      setSelectedHour,
      selectedMinute,
      setSelectedMinute,
      selectedPeriod,
      setSelectedPeriod,
      hours,
      minutes,
      isOpen,
      setIsOpen,
      inputValue,
      handleInputChange,
      handleSave,
      handleCancel,
      timeZonesOptions,
      handleBlur,
      is12HourFormat,
      selectedTimeZone,
      setSelectedTimeZone,
      isDisableSelect,
    } = useTimePicker({
      value,
      defaultValue,
      onChange,
      onBlur,
      timeFormat,
      timeIntervals,
      timeZone,
      showTimezoneSelect,
      includeContinent,
    })
    const generatedId = useId()
    const id = propId ?? generatedId

    if (viewMode === 'edition') {
      return (
        <FormGroup>
          {label && (
            <FormLabel htmlFor={id} required={required} disabled={disabled} hasError={!!errors?.length}>
              {label}
            </FormLabel>
          )}
          <BasePickerField
            id={id}
            iconName="Clock"
            isOpen={isOpen}
            name={name}
            disabled={disabled}
            required={required}
            value={inputValue}
            setIsOpen={setIsOpen}
            onInputChange={handleInputChange}
            ref={ref}
            placeholder={placeholder}
            handleBlur={handleBlur}
            inputProps={{
              ...inputProps,
              onKeyDown: handleKeyDown,
            }}
            className={String.raw`
            [&.base-picker\\_\\_popover]:pr-3
            [&.base-picker\\_\\_popover]:pb-4
            [&.base-picker\\_\\_popover]:pl-4
            [&.base-picker\\_\\_popover]:pt-3
            ${className}
          `}
          >
            <TimePickerContent
              selectedHour={selectedHour}
              selectedMinute={selectedMinute}
              selectedPeriod={selectedPeriod}
              setSelectedHour={setSelectedHour}
              setSelectedMinute={setSelectedMinute}
              setSelectedPeriod={setSelectedPeriod}
              hours={hours}
              minutes={minutes}
              onSave={handleSave}
              onCancel={handleCancel}
              timeZonesOptions={timeZonesOptions}
              selectProps={selectProps}
              is12HourFormat={is12HourFormat}
              isDisableSelect={isDisableSelect}
              selectedTimeZone={selectedTimeZone as string}
              setSelectedTimeZone={setSelectedTimeZone}
              timeZone={timeZone}
            />
          </BasePickerField>
          {description && <FormDescription>{description}</FormDescription>}
          {warnings && <FormMessageList messages={warnings} type="warning" />}
          {errors && <FormMessageList messages={errors} type="error" />}
        </FormGroup>
      )
    }
    return (
      <TextInputDiff
        value={value ?? defaultValue ?? ''}
        viewMode={viewMode}
        diffMode="sentences"
        baseValue={baseValue ?? ''}
        label={label}
        required={required}
        data-testid="time-picker-diff"
      />
    )
  }
)

TimePicker.displayName = 'TimePicker'

export { TimePicker, type TimePickerProps }
