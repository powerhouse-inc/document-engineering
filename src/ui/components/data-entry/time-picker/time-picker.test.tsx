import { screen } from '@testing-library/react'
import { userEvent } from '@testing-library/user-event'
import { render } from '@testing-library/react'
import { TimePicker } from './time-picker.js'
import { vi } from 'vitest'

describe('TimePicker', () => {
  beforeEach(() => {
    // Mock the timezone to be consistent across all tests
    vi.spyOn(Intl, 'DateTimeFormat').mockImplementation(
      () =>
        ({
          resolvedOptions: () => ({
            timeZone: 'America/New_York',
          }),
        }) as any
    )
  })

  afterEach(() => {
    vi.restoreAllMocks()
  })

  it('should render with basic props', () => {
    const { container } = render(
      <TimePicker label="Test Label" name="test-time" id="test-id" required disabled={false} />
    )
    expect(container).toMatchSnapshot()
  })

  it('should display the label when provided', () => {
    const labelText = 'Test Label'
    render(<TimePicker name="test-time" label={labelText} />)
    expect(screen.getByText(labelText)).toBeInTheDocument()
  })

  it('should not render the label when label prop is not provided', () => {
    render(<TimePicker name="test-time" />)
    expect(screen.queryByText('Test Label')).not.toBeInTheDocument()
  })

  it('should mark the label as required when required prop is true', () => {
    render(<TimePicker name="test-time" label="Test Label" required />)
    const label = screen.getByText('Test Label')
    const asterisk = screen.getByText('*')
    expect(label).toBeInTheDocument()
    expect(asterisk).toBeInTheDocument()
  })

  it('should mark the label as disabled when disabled prop is true', () => {
    render(<TimePicker name="test-time" label="Test Label" disabled />)
    const label = screen.getByText('Test Label')
    expect(label).toHaveClass('cursor-not-allowed')
    expect(label).toHaveClass('text-gray-700')
  })

  it('should handle timeIntervals prop correctly with 15-minute intervals', async () => {
    const user = userEvent.setup()
    render(<TimePicker name="test-time" label="Test Label" timeIntervals={15} />)

    const clockButton = screen.getByRole('button')
    await user.click(clockButton)

    const popoverContent = await screen.findByRole('dialog')
    expect(popoverContent).toBeVisible()

    const minutes = ['00', '15', '30', '45']
    for (const minute of minutes) {
      const minuteElements = await screen.findAllByText(minute)
      expect(minuteElements.length).toBeGreaterThan(0)
      expect(minuteElements[0]).toBeInTheDocument()
    }

    ;['13', '20', '25', '35', '40', '50', '55'].forEach(minute => {
      expect(screen.queryByText(minute)).not.toBeInTheDocument()
    })
  })

  it('should handle timeIntervals prop correctly with 30-minute intervals', async () => {
    const user = userEvent.setup()
    render(<TimePicker name="test-time" label="Test Label" timeIntervals={30} />)

    const clockButton = screen.getByRole('button')
    await user.click(clockButton)

    const popoverContent = await screen.findByRole('dialog')
    expect(popoverContent).toBeVisible()

    const minutes = ['00', '30']
    for (const minute of minutes) {
      const minuteElements = await screen.findAllByText(minute)
      expect(minuteElements.length).toBeGreaterThan(0)
      expect(minuteElements[0]).toBeInTheDocument()
    }

    ;['15', '45'].forEach(minute => {
      expect(screen.queryByText(minute)).not.toBeInTheDocument()
    })
  })

  it('should handle timezone selection when timeZone prop is provided', async () => {
    const user = userEvent.setup()
    const timeZoneValue = 'America/New_York'

    render(<TimePicker name="test-time" label="Test Label" timeZone={timeZoneValue} showTimezoneSelect />)

    const clockButton = screen.getByRole('button')
    await user.click(clockButton)

    const popoverContent = await screen.findByRole('dialog')
    expect(popoverContent).toBeVisible()

    const select = await screen.findByRole('combobox')
    expect(select).toBeDisabled()

    const timezoneText = screen.getByText(content => {
      return content.includes('New York') && content.includes('(')
    })
    expect(timezoneText).toBeInTheDocument()
  })

  it('should display description when provided', () => {
    const description = 'Test description'
    render(<TimePicker name="test-time" label="Test Label" description={description} />)
    expect(screen.getByText(description)).toBeInTheDocument()
  })

  it('should display error messages when provided', () => {
    const errors = ['Error 1', 'Error 2']
    render(<TimePicker name="test-time" label="Test Label" errors={errors} />)
    errors.forEach(error => {
      expect(screen.getByText(error)).toBeInTheDocument()
    })
  })

  it('should display warning messages when provided', () => {
    const warnings = ['Warning 1', 'Warning 2']
    render(<TimePicker name="test-time" label="Test Label" warnings={warnings} />)
    warnings.forEach(warning => {
      expect(screen.getByText(warning)).toBeInTheDocument()
    })
  })

  it('should handle different time formats', async () => {
    const user = userEvent.setup()
    render(<TimePicker name="test-time" label="Test Label" timeFormat="HH:mm" />)

    const clockButton = screen.getByRole('button')
    await user.click(clockButton)

    const popoverContent = await screen.findByRole('dialog')
    expect(popoverContent).toBeVisible()

    // Verify 24-hour format is used
    const hours = [
      '00',
      '01',
      '02',
      '03',
      '04',
      '05',
      '06',
      '07',
      '08',
      '09',
      '10',
      '11',
      '12',
      '13',
      '14',
      '15',
      '16',
      '17',
      '18',
      '19',
      '20',
      '21',
      '22',
      '23',
    ]
    for (const hour of hours) {
      const hourElements = await screen.findAllByText(hour)
      expect(hourElements.length).toBeGreaterThan(0)
    }
  })

  it('should show timezone select when showTimezoneSelect is true', async () => {
    const user = userEvent.setup()
    render(<TimePicker name="test-time" label="Test Label" showTimezoneSelect />)

    const clockButton = screen.getByRole('button')
    await user.click(clockButton)

    const popoverContent = await screen.findByRole('dialog')
    expect(popoverContent).toBeVisible()

    const select = await screen.findByRole('combobox')
    expect(select).toBeInTheDocument()
    expect(select).not.toBeDisabled()
  })

  it('should include continent in timezone display when includeContinent is true', async () => {
    const user = userEvent.setup()
    const timeZoneValue = 'America/New_York'

    render(
      <TimePicker name="test-time" label="Test Label" timeZone={timeZoneValue} showTimezoneSelect includeContinent />
    )

    const clockButton = screen.getByRole('button')
    await user.click(clockButton)

    const popoverContent = await screen.findByRole('dialog')
    expect(popoverContent).toBeVisible()

    const timezoneText = screen.getByText(content => {
      return content.includes('America') && content.includes('New York')
    })
    expect(timezoneText).toBeInTheDocument()
  })

  it('should not include continent in timezone display when includeContinent is false', async () => {
    const user = userEvent.setup()
    const timeZoneValue = 'America/New_York'

    render(
      <TimePicker
        name="test-time"
        label="Test Label"
        timeZone={timeZoneValue}
        showTimezoneSelect
        includeContinent={false}
      />
    )

    const clockButton = screen.getByRole('button')
    await user.click(clockButton)

    const popoverContent = await screen.findByRole('dialog')
    expect(popoverContent).toBeVisible()

    const timezoneText = screen.getByText(content => {
      return content.includes('New York') && !content.includes('America')
    })
    expect(timezoneText).toBeInTheDocument()
  })

  it('should not show timezone select when showTimezoneSelect is false', async () => {
    const user = userEvent.setup()
    // Mock the timezone to be Montevideo
    vi.spyOn(Intl, 'DateTimeFormat').mockImplementation(
      () =>
        ({
          resolvedOptions: () => ({
            timeZone: 'America/Montevideo',
          }),
          formatToParts: () => [{ type: 'timeZoneName', value: 'GMT-03:00' }],
        }) as any
    )

    render(<TimePicker name="test-time" label="Test Label" timeZone="America/Montevideo" showTimezoneSelect={false} />)

    const clockButton = screen.getByRole('button')
    await user.click(clockButton)

    const popoverContent = await screen.findByRole('dialog')
    expect(popoverContent).toBeVisible()

    const select = screen.getByRole('combobox')
    expect(select).toBeInTheDocument()
    expect(select).toHaveTextContent('(UTC-03) Montevideo')
  })
})
