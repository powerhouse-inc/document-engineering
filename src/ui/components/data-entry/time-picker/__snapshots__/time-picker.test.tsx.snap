// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`TimePicker > should render with basic props 1`] = `
<div>
  <div
    class="flex flex-col gap-2"
  >
    <label
      class="inline-flex items-center text-sm font-medium leading-4 text-gray-900 dark:text-gray-50 mb-[3px]"
      for="test-id"
    >
      Test Label
      <span
        class="ml-1 text-gray-800 group-hover:text-gray-900 dark:text-gray-400 dark:group-hover:text-slate-50"
      >
        *
      </span>
    </label>
    <div
      class="flex flex-col space-y-2"
    >
      <div
        class="flex w-full rounded-md text-sm focus-within:ring-ring dark:ring-charcoal-300 ring-gray-900 focus-within:ring-1 focus-within:ring-offset-0 dark:border-charcoal-700 dark:bg-charcoal-900 border border-gray-300 bg-white focus:[&_.input-field]:bg-transparent dark:focus-within:hover:bg-charcoal-900 focus-within:hover:cursor-default focus-within:hover:bg-white base-picker__input [&.base-picker\\\\_\\\\_popover]:pr-3 [&.base-picker\\\\_\\\\_popover]:pb-4 [&.base-picker\\\\_\\\\_popover]:pl-4 [&.base-picker\\\\_\\\\_popover]:pt-3 undefined"
      >
        <button
          aria-controls="radix-:r0:"
          aria-expanded="false"
          aria-haspopup="dialog"
          class="inline-flex items-center justify-center gap-2 whitespace-nowrap rounded-md text-sm font-medium transition-colors focus-visible:outline-none disabled:pointer-events-none [&_svg]:pointer-events-none [&_svg]:size-4 [&_svg]:shrink-0 hover:bg-accent hover:text-accent-foreground pl-3 pr-3 z-1 focus:text-selected-foreground focus:bg-none button-ghost focus:outline-none focus-visible:ring-1 focus-visible:ring-offset-0 focus-visible:ring-gray-900"
          data-state="closed"
          type="button"
        >
          <svg
            class="hover:none text-gray-700 dark:text-gray-50"
            data-testid="icon-fallback"
            fill="none"
            height="16"
            style="width: 16px; height: 16px;"
            viewBox="0 0 16 16"
            width="16"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              clip-rule="evenodd"
              d="M8.00022 0.967133C4.11618 0.967133 0.967529 4.11578 0.967529 7.99982C0.967529 11.8838 4.11617 15.0325 8.00022 15.0325C11.8842 15.0325 15.0329 11.8838 15.0329 7.99982C15.0329 4.11578 11.8842 0.967133 8.00022 0.967133ZM1.91753 7.99982C1.91753 4.64045 4.64085 1.91713 8.00022 1.91713C11.3596 1.91713 14.0829 4.64045 14.0829 7.99982C14.0829 11.3592 11.3596 14.0825 8.00022 14.0825C4.64085 14.0825 1.91753 11.3592 1.91753 7.99982ZM8.50012 4.8C8.50012 4.52386 8.27626 4.3 8.00012 4.3C7.72398 4.3 7.50012 4.52386 7.50012 4.8V8C7.50012 8.13261 7.5528 8.25979 7.64657 8.35355L9.7799 10.4869C9.97516 10.6821 10.2917 10.6821 10.487 10.4869C10.6823 10.2916 10.6823 9.97504 10.487 9.77978L8.50012 7.79289V4.8Z"
              fill="currentColor"
              fill-rule="evenodd"
            />
          </svg>
        </button>
        <input
          class="flex h-9 rounded-md text-sm font-normal leading-5 text-gray-900 dark:text-gray-50 dark:border-charcoal-700 dark:bg-charcoal-900 border border-gray-300 bg-white px-3 py-2 font-sans placeholder:text-gray-500 dark:placeholder:text-gray-600 focus-visible:outline-none focus-visible:ring-gray-900 focus-visible:ring-offset-white dark:focus-visible:ring-charcoal-300 dark:focus-visible:ring-offset-charcoal-900 dark:focus:bg-charcoal-900 disabled:cursor-not-allowed disabled:border-gray-300 disabled:bg-gray-50 disabled:text-gray-700 disabled:dark:border-charcoal-800 disabled:dark:bg-charcoal-900 disabled:dark:text-gray-300 w-full rounded-l-none border-none text-right placeholder:text-right input-field focus:bg-white focus-visible:ring-0 focus-visible:ring-offset-0"
          name="test-time"
          required=""
          value=""
        />
      </div>
    </div>
  </div>
</div>
`;
